File: python/day01.py
def get_measurements():
    with open("inputs/day01.txt") as f:
        return [int(x) for x in f.readlines()]


def compare_pairs(L):
    return sum(L[i] < L[i+1] for i in range(len(L)-1))


def part1():
    return compare_pairs(get_measurements())


def part2():
    measurements = get_measurements()
    sums = [sum(measurements[i:i+3]) for i in range(len(measurements)-2)]
    return compare_pairs(sums)


if __name__ == "__main__":
    print(part1())
    print(part2())

File: ocaml/day01.ml
open Core


let measurements =
  In_channel.read_lines "inputs/day01.txt"
  |> List.map ~f:int_of_string

let part1 =
  let rec find_increases measures n =
    match measures with
      | prev :: next :: others when next > prev -> find_increases (next :: others) (n + 1)
      | _ :: next :: others -> find_increases (next :: others) n
      | _ -> n in
  find_increases measurements 0

let part2 =
  let rec find_sum_increases measures n =
    match measures with
      | a :: b :: c :: d :: others 
        when (a + b + c) < (b + c + d) 
        -> find_sum_increases (b :: c :: d :: others) (n + 1)
      | _ :: next :: others -> find_sum_increases (next :: others) n
      | _ -> n in
  find_sum_increases measurements 0

let _ =
  print_endline (string_of_int part1);
  print_endline (string_of_int part2)
